From 07eba9f83de1a0630cf99ac25080a67701912092 Mon Sep 17 00:00:00 2001
From: Laurent Pinchart <laurent.pinchart+renesas@ideasonboard.com>
Date: Tue, 5 Mar 2013 00:48:56 +0100
Subject: [PATCH 0982/1083] drm/rcar-du: Fix plane index wrap-around for
 multi-planar overlays

If the first hardware plane of a multi-planar overlay is plane 7 the
second hardware plane will be plane 0. Handle the wrap-around when
configuring plane priorities.

Signed-off-by: Laurent Pinchart <laurent.pinchart+renesas@ideasonboard.com>
(cherry picked from commit 423be341e4c7171b96edfc8e283bcd5bdfc00c8a)

Signed-off-by: Simon Horman <horms+renesas@verge.net.au>
(cherry picked from commit 1dfe53bfd367a07532c417707ca334a9132447c9)

Signed-off-by: Ryo Kataoka <ryo.kataoka.wt@renesas.com>
---
 drivers/gpu/drm/rcar-du/rcar_du_crtc.c | 7 +++++--
 1 file changed, 5 insertions(+), 2 deletions(-)

diff --git a/drivers/gpu/drm/rcar-du/rcar_du_crtc.c b/drivers/gpu/drm/rcar-du/rcar_du_crtc.c
index 22b87ec..ca80939 100644
--- a/drivers/gpu/drm/rcar-du/rcar_du_crtc.c
+++ b/drivers/gpu/drm/rcar-du/rcar_du_crtc.c
@@ -142,13 +142,16 @@ void rcar_du_crtc_update_planes(struct drm_crtc *crtc)
 
 	for (i = 0; i < num_planes; ++i) {
 		struct rcar_du_plane *plane = planes[i];
+		unsigned int index = plane->hwindex;
 
 		prio -= 4;
-		dspr |= (plane->hwindex + 1) << prio;
+		dspr |= (index + 1) << prio;
 
 		if (plane->format->planes == 2) {
+			index = (index + 1) % 8;
+
 			prio -= 4;
-			dspr |= (plane->hwindex + 2) << prio;
+			dspr |= (index + 1) << prio;
 		}
 	}
 
-- 
1.8.3.2

